File: wrappers/android/src/main/kotlin/com/printerdriver/PrintJob.kt
package com.printerdriver

import com.printerdriver.internal.NativeBridge
import com.printerdriver.internal.NativeJobEventCallback
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.transformWhile
import kotlinx.coroutines.isActive
import kotlinx.coroutines.withContext

/**
 * One submitted job (docs/api.md §2): event stream + queryable state, durable across
 * restarts. Two [PrintJob] instances wrapping the same underlying pd_job* compare
 * equal ([equals]/[hashCode] are handle-based) -- deliberately, since pd.h's own
 * pointer-equality dedupe guarantee ("the same underlying job always maps to the same
 * pd_job pointer, so idempotency-key dedupe is visible as pointer equality") is
 * exactly the property a re-wrap (e.g. two [Printer.print] calls with the same [key],
 * or a [PrinterDriver.findJob] after a fresh submit) should preserve at this layer
 * too, without needing a Kotlin-side interning cache to do it.
 */
class PrintJob internal constructor(
    private val driver: PrinterDriver,
    internal val handle: Long
) {
    val id: String get() { driver.checkOpen(); return NativeBridge.jobId(handle) }
    val key: String get() { driver.checkOpen(); return NativeBridge.jobKey(handle) }

    /** The receipt verification identifier this attempt printed under -- the four
     *  characters the ticket carries as `V:` (docs/api.md §14). `null` until the job
     *  reaches a worker, and `null` for good on a printer whose fence is not GS(H):
     *  the identifier *is* the wire token. Resolve one with [PrinterDriver.jobByToken]. */
    val printToken: String? get() = NativeBridge.jobPrintToken(handle).ifEmpty { null }

    /** The identifier the job's cut fence carried. Same rules as [printToken]. */
    val cutToken: String? get() = NativeBridge.jobCutToken(handle).ifEmpty { null }

    /** 1 on first submission; incremented by each [Printer.forceReprint]. */
    val attempt: Int get() { driver.checkOpen(); return NativeBridge.jobAttempt(handle) }

    val state: JobState get() { driver.checkOpen(); return JobState.fromRaw(NativeBridge.jobCurrentState(handle)) }
    val confidence: ConfidenceLevel get() { driver.checkOpen(); return ConfidenceLevel.fromRaw(NativeBridge.jobConfidence(handle)) }
    val isTerminal: Boolean get() { driver.checkOpen(); return NativeBridge.jobIsTerminal(handle) }

    /**
     * The job's full event history, replayed, then live -- ends with a terminal event,
     * always (pd.h). Collecting this twice subscribes twice: pd_subscribe_job's replay
     * is per-call, so each collector independently gets the complete ordered history
     * from the start, exactly as if it had subscribed first. That is a deliberate
     * consequence of this being backed by [callbackFlow] (cold) rather than a shared/
     * hot flow -- see README.md "Threading contract" for why this wrapper does not
     * multicast a single native subscription across collectors.
     *
     * The native callback (see printerdriver_jni.cpp) only ever calls `trySend` on this
     * flow's channel; it never calls back into any native accessor. [isTerminal] is
     * checked below, in the [transformWhile] operator, specifically so that check runs
     * on the *collector's* coroutine -- never inside the native trampoline's call stack
     * -- because pd.h forbids a
     * callback from calling back into any pd_* function on the same driver ("the
     * worker thread it is running on is the thread that would have to service that
     * call"). This is not a cosmetic ordering choice; doing the isTerminal check inside
     * the callback itself would violate that contract.
     */
    val events: Flow<JobEvent> = callbackFlow {
        driver.checkOpen()
        val callback = NativeJobEventCallback { state, confidence, hasReason, reason, monotonicMs ->
            trySend(JobEvent.fromRaw(state, confidence, hasReason, reason, monotonicMs))
        }
        NativeBridge.subscribeJob(driver.handle, handle, callback)
        awaitClose {
            // No pd_unsubscribe_job exists (see NativeBridge.subscribeJob's doc). The
            // GlobalRef this subscription holds on the native side is released when
            // the owning PrinterDriver is closed, not when collection stops -- a
            // cancelled collector simply stops reading from this channel; the native
            // callback keeps firing (harmlessly, into a closed channel) until then.
        }
    }.transformWhile { event ->
        emit(event)
        !isTerminal
    }

    /**
     * Waits for the terminal result (pd_job_await). Implemented as a bounded poll
     * rather than a single indefinite native wait so that cancelling the calling
     * coroutine is noticed within [RESULT_POLL_INTERVAL_MS] instead of not at all --
     * pd_job_await is a blocking native call with no cancellation hook of its own, so
     * cooperative cancellation has to come from never blocking on it for longer than
     * one poll interval at a time.
     */
    suspend fun result(): JobResult = withContext(Dispatchers.IO) {
        while (isActive) {
            driver.checkOpen()
            val raw = NativeBridge.jobAwait(driver.handle, handle, RESULT_POLL_INTERVAL_MS)
            if (raw != null) {
                // The method string is fetched only once the result exists, so the
                // extra call is on the settled path and never on the polling one.
                return@withContext JobResult.fromRaw(
                    raw[0], raw[1], raw[2], raw[3], raw[4],
                    NativeBridge.jobMethod(driver.handle, handle)
                )
            }
        }
        throw CancellationException("PrintJob.result() was cancelled")
    }

    override fun equals(other: Any?): Boolean = other is PrintJob && other.handle == handle
    override fun hashCode(): Int = handle.hashCode()
    override fun toString(): String = "PrintJob(id=$id, key=$key, state=$state)"

    private companion object {
        const val RESULT_POLL_INTERVAL_MS = 250
    }
}

File: wrappers/dotnet/PrinterDriver/PrintJob.cs
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Threading.Channels;

namespace PrinterDriver;

/// <summary>
/// One submitted job. Owned by its <see cref="PrinterDriver"/>; never disposed
/// individually.
/// </summary>
public sealed class PrintJob
{
    private readonly PrinterDriver _driver;
    private readonly nint _handle;
    private readonly object _gate = new();
    private readonly List<Channel<JobEvent>> _channels = [];
    private bool _subscribed;
    private bool _driverDisposed;

    internal PrintJob(PrinterDriver driver, nint handle)
    {
        _driver = driver;
        _handle = handle;
        Id = NativeMethods.ReadUtf8(NativeMethods.pd_job_id(handle));
        Key = NativeMethods.ReadUtf8(NativeMethods.pd_job_key(handle));
    }

    /// <summary>The job's own identifier, stable for the life of the driver.</summary>
    public string Id { get; }

    /// <summary>The idempotency key it was submitted under.</summary>
    public string Key { get; }

    /// <summary>Which attempt this is; a force-reprint increments it.</summary>
    public uint Attempt => NativeMethods.pd_job_attempt(_handle);

    /// <summary>Where the job is right now.</summary>
    public JobState CurrentState => (JobState)NativeMethods.pd_job_current_state(_handle);

    /// <summary>What the current state rests on.</summary>
    public ConfidenceLevel Confidence => (ConfidenceLevel)NativeMethods.pd_job_confidence(_handle);

    /// <summary>True once the job has reached a terminal state.</summary>
    public bool IsTerminal => NativeMethods.pd_job_is_terminal(_handle) != 0;

    /// <summary>
    /// The verification identifier this job's ticket printed as <c>V:</c> — the wire token
    /// its print fence carried, promoted to paper (docs/api.md §14).
    /// </summary>
    /// <remarks>
    /// Null on a printer whose completion mechanism is not <c>GS ( H</c> (there is no wire
    /// token to promote) and until the job reaches a worker. Once non-null it never
    /// changes, and <see cref="PrinterDriver.JobByToken(string)"/> resolves it back to this
    /// job, including after a restart.
    /// </remarks>
    public string? PrintToken
    {
        get
        {
            var token = NativeMethods.ReadUtf8(NativeMethods.pd_job_print_token(_handle));
            return token.Length == 0 ? null : token;
        }
    }

    /// <summary>The identifier this job's cut fence carried. Same rules as
    /// <see cref="PrintToken"/>.</summary>
    public string? CutToken
    {
        get
        {
            var token = NativeMethods.ReadUtf8(NativeMethods.pd_job_cut_token(_handle));
            return token.Length == 0 ? null : token;
        }
    }

    /// <summary>
    /// The job's full event stream: every event recorded so far, replayed on the calling
    /// thread, then the rest as the printer's worker thread produces them. The last event
    /// a job ever emits is always a terminal one, which is what ends the enumeration.
    /// </summary>
    /// <param name="cancellationToken">Stops the enumeration; the job keeps going.</param>
    /// <returns>The events, in order.</returns>
    public IAsyncEnumerable<JobEvent> EventsAsync(
        CancellationToken cancellationToken = default)
    {
        // Subscribe eagerly, before the iterator is ever advanced: the body of an
        // `async IAsyncEnumerable` method does not run until the first MoveNextAsync, and
        // anything the worker thread emitted in between would be gone.
        var reader = Subscribe();
        return Enumerate(reader, cancellationToken);
    }

    private static async IAsyncEnumerable<JobEvent> Enumerate(
        ChannelReader<JobEvent> reader,
        [EnumeratorCancellation] CancellationToken cancellationToken)
    {
        await foreach (var jobEvent in reader.ReadAllAsync(cancellationToken)
                           .ConfigureAwait(false))
        {
            yield return jobEvent;
            if (jobEvent.IsTerminal)
            {
                yield break;
            }
        }
    }

    /// <summary>
    /// Waits for the terminal result.
    /// </summary>
    /// <remarks>
    /// Implemented as a short polling loop over <c>pd_job_await</c> rather than one
    /// indefinite native call, specifically so that a cancelled token is noticed within
    /// one poll interval: <c>pd_job_await</c> itself has no cancellation hook, and a
    /// thread parked inside it cannot be interrupted.
    /// </remarks>
    /// <param name="cancellationToken">Stops waiting; the job itself keeps going.</param>
    /// <returns>Done, Failed or Unknown — never a boolean.</returns>
    public async Task<JobResult> GetResultAsync(CancellationToken cancellationToken = default)
    {
        const uint pollMs = 100;
        while (true)
        {
            cancellationToken.ThrowIfCancellationRequested();
            var result = await Task.Run(() =>
            {
                var ready = NativeMethods.pd_job_await(_driver.Handle, _handle, pollMs,
                                                       out var native);
                return ready != 0 ? JobResult.FromNative(in native) : null;
            }, cancellationToken).ConfigureAwait(false);
            if (result is not null)
            {
                return result;
            }
        }
    }

    /// <summary>
    /// Waits for the terminal result, blocking the calling thread.
    /// </summary>
    /// <param name="timeoutMs">How long to wait; 0 waits indefinitely.</param>
    /// <returns>The result, or null on timeout.</returns>
    public JobResult? WaitForResult(uint timeoutMs = 0)
    {
        var ready = NativeMethods.pd_job_await(_driver.Handle, _handle, timeoutMs, out var native);
        return ready != 0 ? JobResult.FromNative(in native) : null;
    }

    internal nint Handle => _handle;

    internal void OnDriverDisposed()
    {
        lock (_gate)
        {
            _driverDisposed = true;
            foreach (var channel in _channels)
            {
                channel.Writer.TryComplete();
            }
            _channels.Clear();
        }
    }

    private ChannelReader<JobEvent> Subscribe()
    {
        var channel = Channel.CreateUnbounded<JobEvent>(new UnboundedChannelOptions
        {
            SingleWriter = false,
            SingleReader = true,
        });
        lock (_gate)
        {
            if (_driverDisposed)
            {
                channel.Writer.TryComplete();
                return channel.Reader;
            }
            _channels.Add(channel);
            if (_subscribed)
            {
                return channel.Reader;
            }
            // First subscriber wins the one native subscription. pd_subscribe_job replays
            // the recorded history synchronously on this thread before returning, so the
            // channel is already primed by the time the caller starts reading and no
            // event can be missed between subscribing and enumerating.
            var context = _driver.Roots.Root(this);
            var callback = new NativeMethods.JobEventCallback(Trampoline);
            _driver.Roots.Root(callback);
            _subscribed = true;
            NativeMethods.pd_subscribe_job(_driver.Handle, _handle, callback,
                                           GCHandle.ToIntPtr(context));
            return channel.Reader;
        }
    }

    private static void Trampoline(nint _, PdJobEvent jobEvent, nint context)
    {
        var job = CallbackRoots.FromContext<PrintJob>(context);
        if (job is null)
        {
            return;
        }
        job.Publish(JobEvent.FromNative(jobEvent));
    }

    private void Publish(JobEvent jobEvent)
    {
        Channel<JobEvent>[] channels;
        lock (_gate)
        {
            channels = [.. _channels];
        }
        foreach (var channel in channels)
        {
            channel.Writer.TryWrite(jobEvent);
            if (jobEvent.IsTerminal)
            {
                channel.Writer.TryComplete();
            }
        }
    }
}
